### PR TITLE
Cryoemio

### PR DESCRIPTION
**Pull request by thisTyler**
Added cryoemio.py and corresponding tests. This doc consists of helper functions originally designed for the TEM wrapper.

### CODE REVIEW

**Comment by thisTyler** [comment]
@geoffwoollard This PR consists of helper functions for the TEM wrapper and corresponding tests - all is working and we already have use-cases for these functions in one of our TEM wrapper branches. Good to merge?

File: ioSPI/cryoemio.py [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+    with h5py.File(h5_file, "w") as file:
+        recursively_save_dict_contents_to_group(file, "/", dic)
+
+
+def recursively_save_dict_contents_to_group(h5file, path, dic):
+    """Recursively save dictionary contents to group.
+
+    Parameters
+    ----------
+    h5file : File
```
Comment (geoffwoollard) [inline ioSPI/cryoemio.py:46]: what type is File?

File: ioSPI/cryoemio.py [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+
+
+def fill_parameters_dictionary(
+    yaml_file, mrc_file, pdb_file, crd_file, log_file, dose=None, noise=None
+):
+    """Return parameter dictionary with settings for simulation.
+
+    Parameters
+    ----------
+    yaml_file : str
```
Comment (geoffwoollard) [inline ioSPI/cryoemio.py:71]: eventually there will be many yaml files with different entries. a more informative name would be better.

File: ioSPI/cryoemio.py [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+    dic = {"simulation": {}}
+    if "miscellaneous" in parameters:
+        if "seed" in parameters["miscellaneous"]:
+            dic["simulation"]["seed"] = parameters["miscellaneous"]["seed"]
+        else:
+            random.seed()
+            dic["simulation"]["seed"] = random.randint(0, int(1e10))
+    else:
+        random.seed()
+        dic["simulation"]["seed"] = random.randint(0, int(1e10))
```
Comment (geoffwoollard) [inline ioSPI/cryoemio.py:149]: could check for dic["simulation"]["seed"], and fill if empty, so don't have to have two else statements

File: ioSPI/cryoemio.py [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+    else:
+        dic["particle"]["map_file_re_out"] = None
+    dic["particleset"] = {}
+    dic["particleset"]["name"] = parameters["molecular_model"]["particle_name"]
+    dic["particleset"]["crd_file"] = crd_file
+    dic["beam"] = {}
+    dic["beam"]["voltage"] = parameters["beam_parameters"]["voltage"]  # voltage in kV
+    dic["beam"]["spread"] = parameters["beam_parameters"][
+        "energy_spread"
+    ]  # energy spread in V
```
Comment (geoffwoollard) [inline ioSPI/cryoemio.py:178]: put these comments in the code somehow, by putting units in variable names. comments are not good to maintain

File: ioSPI/cryoemio.py [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+    if dose is not None:
+        dic["beam"]["dose_per_im"] = dose
+    else:
+        dic["beam"]["dose_per_im"] = parameters["beam_parameters"][
+            "electron_dose"
+        ]  # dose per image in e/nm**2
```
Comment (geoffwoollard) [inline ioSPI/cryoemio.py:184]: put units in variable name

File: ioSPI/cryoemio.py [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+    dic["beam"]["dose_sd"] = parameters["beam_parameters"][
+        "electron_dose_std"
+    ]  # standard deviation of dose per image
```
Comment (geoffwoollard) [inline ioSPI/cryoemio.py:187]: put units in variable name

File: ioSPI/cryoemio.py [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+    dic["optics"] = {}
+
+    dic["optics"]["magnification"] = parameters["optics_parameters"][
+        "magnification"
+    ]  # magnification
+    dic["optics"]["cs"] = parameters["optics_parameters"][
+        "spherical_aberration"
+    ]  # spherical aberration in mm
```
Comment (geoffwoollard) [inline ioSPI/cryoemio.py:195]: put units in variable name

File: ioSPI/cryoemio.py [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+    dic["optics"]["cc"] = parameters["optics_parameters"][
+        "chromatic_aberration"
+    ]  # chromatic aberration in mm
```
Comment (geoffwoollard) [inline ioSPI/cryoemio.py:198]: put units in variable name

File: ioSPI/cryoemio.py [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+    dic["optics"]["aperture"] = parameters["optics_parameters"][
+        "aperture_diameter"
+    ]  # diameter in um of aperture in back focal plane
```
Comment (geoffwoollard) [inline ioSPI/cryoemio.py:201]: put units in variable name

File: ioSPI/cryoemio.py [hunk 10]
```diff
@@ -0,0 +1,362 @@
...
+    dic["optics"]["focal_length"] = parameters["optics_parameters"][
+        "focal_length"
+    ]  # focal length in mm of primary lens
```
Comment (geoffwoollard) [inline ioSPI/cryoemio.py:204]: put units in variable name

File: ioSPI/cryoemio.py [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+    dic["optics"]["cond_ap_angle"] = parameters["optics_parameters"][
+        "aperture_angle"
+    ]  # aperture angle in mrad of the beam furnished by the condenser lens
```
Comment (geoffwoollard) [inline ioSPI/cryoemio.py:207]: put units in variable name

File: ioSPI/cryoemio.py [hunk 12]
```diff
@@ -0,0 +1,362 @@
...
+        dic["optics"]["defocus_file_out"] = None
+    dic["detector"]["det_pix_x"] = parameters["detector_parameters"][
+        "detector_Nx"
+    ]  # number of pixels on detector along x axis
+    dic["detector"]["det_pix_y"] = parameters["detector_parameters"][
+        "detector_Ny"
+    ]  # number of pixels on detector along y axis
+    dic["detector"]["pixel_size"] = parameters["detector_parameters"][
+        "detector_pixel_size"
+    ]  # physical pixel size in um
```
Comment (geoffwoollard) [inline ioSPI/cryoemio.py:237]: put units in variable name

File: ioSPI/cryoemio.py [hunk 13]
```diff
@@ -0,0 +1,362 @@
...
+    dic["detector"]["gain"] = parameters["detector_parameters"][
+        "detector_gain"
+    ]  # detector gain: average number of counts per electron
```
Comment (geoffwoollard) [inline ioSPI/cryoemio.py:240]: just name it gain_average_count_per_electron instead of the comment

**Review by geoffwoollard — CHANGES_REQUESTED** [review]
overall good job and test coverage. 

Please rename parameter fields with units so don't have to annotate with comments, and is more readable.

File: ioSPI/cryoemio.py (hunk 1 above)
Comment (thisTyler) [reply ioSPI/cryoemio.py:46]: Updated to hdf5.File for clarity

File: ioSPI/cryoemio.py (hunk 2 above)
Comment (thisTyler) [reply ioSPI/cryoemio.py:71]: updated to input_params_file

File: ioSPI/cryoemio.py (hunk 3 above)
Comment (thisTyler) [reply ioSPI/cryoemio.py:149]: changed to a try/except loop waiting for a KeyError as a more clear way of accomplishing the same task.

**Pull request merged**